nbs/20a_distributed.ipynb

```diff
@@ -642,18 +642,17 @@
     "```python\n",
     "---\n",
     "from fastai.vision.all import *\n",
-    "from fastai.accelerate import *\n",
-    "\n",
-    "set_seed(99, True)\n",
-    "path = untar_data(URLs.PETS)/'images'\n",
-    "dls = ImageDataLoaders.from_name_func(\n",
-    "    path, get_image_files(path), valid_pct=0.2,\n",
-    "    label_func=lambda x: x[0].isupper(), item_tfms=Resize(224))\n",
-    "    \n",
-    "learn = vision_learner(dls, resnet34, metrics=error_rate).to_fp16()\n",
+    "from fastai.distributed import *\n",
     "\n",
     "def train():\n",
-    "    with learn.accelerate_ctx(in_notebook=True):\n",
+    "    set_seed(99, True)\n",
+    "    path = untar_data(URLs.PETS)/'images'\n",
+    "    dls = ImageDataLoaders.from_name_func(\n",
+    "        path, get_image_files(path), valid_pct=0.2,\n",
+    "        label_func=lambda x: x[0].isupper(), item_tfms=Resize(224))\n",
+    "    \n",
+    "    learn = vision_learner(dls, resnet34, metrics=error_rate).to_fp16()\n",
```

Comment:
If you don't wrap this line in a [rank0_first](https://docs.fast.ai/distributed.html#rank0_first), then you'll spawn 3 downloads of the model.

But if you _just_ wrap this line in a `rank0_first`, like this `learn = rank0_first(lambda: vision_learner(dls, resnet101, metrics=error_rate).to_fp16())`, then you'll get an error:

> RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cuda:1!

So something else needs to be done. I haven't tracked it down yet.

Reply:
Ok. I've tracked it down to [here](https://github.com/fastai/fastai/blob/master/fastai/data/transforms.py#L372).

```
class Normalize(DisplayedTransform):
    "Normalize/denorm batch of `TensorImage`"
    parameters,order = L('mean', 'std'),99
    def __init__(self, mean=None, std=None, axes=(0,2,3)): store_attr()

    @classmethod
    def from_stats(cls, mean, std, dim=1, ndim=4, cuda=True): return cls(*broadcast_vec(dim, ndim, mean, std, cuda=cuda))

    def setups(self, dl:DataLoader):
        if self.mean is None or self.std is None:
            x,*_ = dl.one_batch()
            self.mean,self.std = x.mean(self.axes, keepdim=True),x.std(self.axes, keepdim=True)+1e-7

    def encodes(self, x:TensorImage): return (x-self.mean) / self.std
    def decodes(self, x:TensorImage):
        f = to_cpu if x.device.type=='cpu' else noop
        return (x*f(self.std) + f(self.mean))

    _docs=dict(encodes="Normalize batch", decodes="Denormalize batch")
```
That setup only happens once, so `self.mean` and `self.std` are going to be on whatever device was used during setup, but `x` is going to be on the current device.

I confirmed it by setting `self.mean` and `self.std` inside the `encodes` function. Now I'm trying to find a way to resolve this without having to recalculate the `mean` and `std` on every call to `encodes`.

Reply:
It's like I'm looking for a way to say "Hey, if this Transform gets distributed, then run this callback."

I can't think of a good general-purpose way to handle this.

It's like there's no good way to do this without complicating DataLoaders/Transformers. I don't want a DataLoader to have to peak inside it's list of transformers and conditionally do stuff to specific ones in the case where the DataLoader gets distributed. And I don't want Transtormers to reach up to their DataLoader to tell them the same.

The way things are now, you just kind of have to "know" whether or not a Transformer is `distrib_ctx`-safe and handle it in a one-off way.

I'm not very familiar with this codebase though. I'd love to get the input of someone who is familiar with it.